clients/drcachesim/tools/invariant_checker.cpp

```diff
@@ -624,17 +640,34 @@ invariant_checker_t::parallel_shard_memref(void *shard_data, const memref_t &mem
                         "Nested kernel context switch traces are not expected");
         shard->pre_context_switch_trace_instr_ = shard->prev_instr_;
         shard->between_kernel_context_switch_markers_ = true;
+#ifdef UNIX
+        shard->signal_stack_depth_at_context_switch_trace_start_ =
+            shard->signal_stack_.size();
+#endif
     }
     if (memref.marker.type == TRACE_TYPE_MARKER &&
         memref.marker.marker_type == TRACE_MARKER_TYPE_CONTEXT_SWITCH_END) {
         report_if_false(shard, shard->between_kernel_context_switch_markers_,
                         "Found kernel context switch trace end without start");
+#ifdef UNIX
+        if (shard->between_kernel_context_switch_markers_) {
+            // Protected by above if-condition to avoid noise from another invariant
+            // error due to a missing start marker.
+            report_if_false(shard,
+                            shard->signal_stack_depth_at_context_switch_trace_start_ ==
+                                static_cast<int>(shard->signal_stack_.size()),
+                            "Context switch trace has extra kernel_event marker");
```

Comment:
Worth another comment announcing this invariant.  Do we need to document it anywhere else?  Not sure what other docs we have on these kernel templates.

Reply:
Would be nice to have a proper DynamoRIO.org page on this. I'll work on that.